### ISSUE TITLE
Regression bug: webcam train fails on Safari iOS

### ISSUE THREAD

**Issue by ivnnv** [tensorflow/tfjs#3084]
#### Describe the problem:
`tfjs-examples/webcam-transfer-learning` demo hosted at:
https://storage.googleapis.com/tfjs-examples/webcam-transfer-learning/dist/index.html 
which seems to be a deploy from this commit from 2019-06-12:
https://github.com/tensorflow/tfjs-examples/tree/fc8646fa87de990a2fc0bab9d1268731186d9f04
Works well on both desktop and mobile browsers. 👌 

But the actual code at https://github.com/tensorflow/tfjs-examples/webcam-transfer-learning (2020-04-16) does NOT work on Safari iOS (it still works well on computer based browsers).
When you hit TRAIN it shows the first batch LOSS number and then, NaN for the rest of batches:
![ac700ac9-ab2f-4aaf-85a2-27942e18d4d7](https://user-images.githubusercontent.com/23552631/79396375-0da7e280-7f7c-11ea-9451-99bc4038e9a8.jpg)
```
[Log] onBatchEnd – 1.6904295682907104 (tfjs-examples.46ef903a.js, line 38188)
[Log] onBatchEnd – Nan (tfjs-examples.46ef903a.js, line 38188)
[Log] onBatchEnd – Nan (tfjs-examples.46ef903a.js, line 38188)
...
```
#### Code to reproduce the bug / link to feature request
I forked the actual code from `/tfjs-examples/webcam-transfer-learning` on:
https://github.com/ivnnv/tfjs-webcam-packman
and made a gh-pages build demo on:
https://ivnnv.github.io/tfjs-webcam-pacman/

**Comment by ivnnv** [comment]
I have done more testing and it seems the bug was introduced in v1.2.8:
The same demo using up to 1.2.7 works (although with a high loss rate 4~6)

**Comment by rthadur** [comment]
@ivnnv do you wish to contribute to fix the issue ?

**Comment by ivnnv** [comment]
> @ivnnv do you wish to contribute to fix the issue ?

@rthadur yeah Id love it as Im trying to build something for mobile (and id like to target all devices if possible) that makes use of this (I'm taking the example app just as the initial proof of concept).

If you knowhow to approach the fix and where Id need to go to fix it and it only needed Javascript knowledge I can try to fix it by myself, but if its something related with the base tensorflow's python package, I wouldnt.

**Comment by rthadur** [comment]
cc @tafsiri @dsmilkov

**Comment by tafsiri** [comment]
Confirmed that this fails to converge on iOS 12 (iPhone 8), but does converge well on Android (Pixel3).

I suspect this training instability on iOS is due to lack of WebGL2 support on iOS. It's pretty complex to debug what might introduce this, so I don't know if there is a single specific thing we can do to fix this, we try and maintain numerical stability on iOS despite the limitations but that can be tricky in a training loop.

Going to leave this open in case any one has any ideas to narrow this down on iOS. But we've seen this before on iOS so i'm not terribly optimistic.

For anyone who does want to do training on iOS I think you would need to experiment with data normalization and the ops used by the model to avoid overflow/underflow or exploding/vanishing gradients.

cc @annxingyuan if you have any ideas here or want to keep track of this.

**Comment by rthadur** [comment]
Verified on latest ios , and also the example is updated to use latest tfjs version. could you please verify once ?
